New ray-based multibeam sonar implementation

woensug-choi opened the pull request:
### Progress
- [x] Plugin 'shell' using the Depth camera @dtdavi1 
- [x] Sonar Beam/ray processing implementation @woensug-choi 

### To-dos (in priority order)
- [x]  Parallelization with NVIDIA GPU CUDA cores @woensug-choi 
- [x]  Fix IFFT and optimization @woensug-choi 
- [x]  Add simpler point scattering model and compare @woensug-choi
- [x]  ROS publisher with sonar data format ([Issue link](https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo/issues/22)) @bsb808 
- [x]  Data visualization with RVIZ @dtdavi1 @woensug-choi 

### Notes on Sonar Beam/ray processing implementation
- Currently, a relatively complex point scattering model is implemented
   - 92 seconds to calculate the model for 512 Beams (114 rays in elevation direction each, [Blueview P900](https://rts.as/wp-content/uploads/2018/09/Teledyne-Blueview-P900-Series-Sonar.pdf))
- IFFT not working correctly (Simple library header found on Github was a fail)
   - It seems the NVIDIA CUDA library has an FFT code of its own.
   - It still has the potential to slow the simulator down considering 10 Hz refresh rates
   - Fast external library should be considered (e.g. [Intel MKL](https://software.intel.com/content/www/us/en/develop/tools/math-kernel-library/fast-fourier-transforms.html),  [fftw](http://www.fftw.org/), [NVIDIA Cuda FFT](https://developer.nvidia.com/cufft), etc.)
- Results obtained using MATLAB to IFFT the results
![image](https://user-images.githubusercontent.com/7955120/97438741-914d7d80-1968-11eb-840f-399842d58c68.png)

dtdavi1 committed:
Created imaging sonar plugin "shell"

The sonar model needs to be implemented, but the plugin compiles and
runs now.  It uses a Gazebo DepthCameraSensor and generates a
depth and normals texture from the depth camera output.

woensug-choi committed:
initial commit for sonar processing (fft needs fix, cuda corenot yet)

woensug-choi committed:
temp

woensug-choi committed:
re-commit

woensug-choi committed:
re-commit

woensug-choi committed:
remove cuda compile optios at CMmakeLists.txt

woensug-choi committed:
ignore vscode files

woensug-choi committed:
fixed unchaning

woensug-choi committed:
cuda tests

woensug-choi committed:
remove -std=c++11 option since cuda setting has CXX. cuda core tested with simple function

woensug-choi committed:
Blazingly flast CUDA core sonar calculation

woensug-choi committed:
FFT added. needs test

woensug-choi committed:
old fft removed

dtdavi1 commented:
@woensug-choi I'm getting a link error when I try to run everything.  I've installed CUDA (cuda, cuda-nvcc-11-1, and nvidia-cuda-toolkit), and everything compiles fine with catkin_make.  When I try to run the sonar_tank_blueview_p900.launch example, I'm getting a "undefined symbol: cufftPlan1d" error.  Is something missing from CMakeLists.txt or the launch file I'm using?

woensug-choi committed:
Wrap up commit (with CPU calculation and calc time measure alive)

woensug-choi committed:
Clean Wrap up commit (without CPU calculation, calculation time measure dead)

woensug-choi commented:
#### Core part of the sonar calculation is done! :+1: 
- Features
   - Physical sonar beam/ray calculation
   - NVIDIA CUDA core GPU parallelization

### Progress
- [x] Plugin 'shell' using the Depth camera @dtdavi1 
- [x] Sonar Beam/ray processing implementation @woensug-choi 

### To-dos (in priority order)
- [x]  Parallelization with NVIDIA GPU CUDA cores @woensug-choi 
- [x]  Fix IFFT and optimization @woensug-choi
- [x]  Add simpler point scattering model and compare @woensug-choi
- [x]  Prallelize and optimize ray summation @woensug-choi
- [x]  ROS publisher with sonar data format ([Issue link](https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo/issues/22)) @bsb808 
- [x]  Data visualization with RVIZ @dtdavi1 @woensug-choi 

### 1. Plugin diagram
![image](https://user-images.githubusercontent.com/7955120/98117980-d421cf00-1eed-11eb-90b5-565705c69909.png)

### 2. How-to install and run
#### 2.1. Install CUDA
**Option 1**. Install CUDA on the host machine (I installed latest version CUDA 11 [download link](https://developer.nvidia.com/cuda-downloads?target_os=Linux&target_arch=x86_64&target_distro=Ubuntu&target_version=1804&target_type=debnetwork))

**Option 2**. Use docker file located at `nps_uw_sensor_gazebo/docker` which includes CUDA installation ([reference](https://github.com/Field-Robotics-Lab/dave/wiki/dave-with-Docker)). The docker file at this PR builds `nps_uw_sensors_nvidia_cuda` docker image instead of `dave_nvidia` from dave project.

- build the docker image
```
cd ~/uuv_ws/src/nps_uw_sensors_gazebo/docker/
./build.bash .
```
- Run the container
```
cd ~/uuv_ws/src/nps_uw_sensors_gazebo/docker/
./run.bash nps_uw_sensors_nvidia_cuda
cd ~/uuv_ws
catkin_make
source devel/setup.bash
```
- Join the container from another terminal
```
cd ~/uuv_ws/src/nps_uw_sensors_gazebo/docker/
./join.bash nps_uw_sensors_nvidia_cuda
source ~/uuv_ws/devel/setup.bash
```

#### 2.2. Quickstart
- Run with follwoing command to test
```bash
roslaunch nps_uw_sensors_gazebo sonar_tank_blueview_p900.launch verbose:=true
```
- Raw data logs are saved at `/tmp/SonarRawData_000001.csv` every 5 frames at default. You can turn it off or change the saving rate at `models/blueeview_p900/model.sdf` using `bool <writeLog>` and `int <writeFrameInterval>`.

### Benchmark
- Computation time (Using Laptop with Docker installation on Linux, Intel i5, GeForce GTX 1050 Ti Mobile)
  - CPU linear (w/o FFT) :  about 200 seconds/frame
  - GPU CUDA (w    FFT) : about 1.5 seconds/frame
    - Ray calculation : about 0.7 seconds/frame
    - Ray summation : about 0.85 seconds/frame
    - FFT calculation : about 0.04 seconds/frame
  - Raw data write time     : about 6 seconds/frame (Writing about 125 MB csv file)
- GPU RAM usage
  - With 4GB RAM on laptop, the CUDA exits with error `not enough ram`. Every other rays are calculated to fit in the limit of the RAM size. Skipping Beam/Ray calculations can be configured at `model.sdf` with `int <beamSkips>` and `int <raySkips>`.
- Results
  - Image drawn using jet colormap of the MATLAB (the script is in `nps_uw_sensors_gazebo/scrips/plotRawData.m` with right path for raw csv file)
  - Taking too much time to summate rays. It seems there are ways to use GPU for summation too. Will try that soon. If done with some more optimizations, the current plugin is expected to have about 2 Hz refresh rate.
![image](https://user-images.githubusercontent.com/7955120/98197454-21dc1d00-1f6a-11eb-9291-f09648c9a561.png)

woensug-choi commented:
> @woensug-choi I'm getting a link error when I try to run everything. I've installed CUDA (cuda, cuda-nvcc-11-1, and nvidia-cuda-toolkit), and everything compiles fine with catkin_make. When I try to run the sonar_tank_blueview_p900.launch example, I'm getting a "undefined symbol: cufftPlan1d" error. Is something missing from CMakeLists.txt or the launch file I'm using?

I had missing library at CMakeLists. I've cleaned everything up again and tested on clean Docker environments. Could you try it again?

woensug-choi committed:
Github static code cheking attempt 1

woensug-choi committed:
Github static code cheking attempt 2

woensug-choi committed:
@woensug-choi Github static code cheking attempt 23

woensug-choi committed:
Github static code cheking attempt 4

woensug-choi committed:
Github static code cheking attempt 5

dtdavi1 commented:
@woensug-choi Everything starts up ok now, but as soon as I move the sonar I get the following error and plugin stops.

Kernel Launch Failed
File: /home/dtdavis/ACS/acs_ros_ws/src/Field-Robotics-Lab/nps_uw_sensors_gazebo/src/sonar_calculation_cuda.cu
Line Number: 312
Reason: an illegal memory access was encountered

It's entirely likely that I installed some of the CUDA libraries incorrectly or have the wrong versions.  Any suggestions?

woensug-choi committed:
intermediate commit. bugs in magnitude of the data. but it works

woensug-choi commented:
@dtdavi1 Sorry for the inconvenience. Could you try with the new commit? illegal access in the cuda kernel is hard to debug.. hope the new commit works.

### Current benchmark results

- **Full / 50 m Range / 512 Beams - 114 Rays**
GPU Sonar Computation Time 66/100 [s]
**CPU Sonar Sum - Correction 45395/100 [s]**
GPU FFT  Calculation  Time 14/100 [s]
(TOTAL) GPU Sonar Frame Calc Time 454 [s] = 7.6 [min]

- **Ray Reduced / 50 m Range / 512 Beams - 11 Rays**
GPU Sonar Computation Time 7/100 [s]
**CPU Sonar Sum - Correction 3103/100 [s]**
GPU FFT  Calculation  Time 4/100 [s]
(TOTAL) GPU Sonar Frame Calc Time 30.3 [s]

- **Ray/Range Reduced / 10 m Range / 512 Beams - 11 Rays**
GPU Sonar Computation Time 1/100 [s]
**CPU Sonar Sum - Correction 481/100 [s]**
GPU FFT  Calculation  Time 1/100 [s]
(TOTAL) GPU Sonar Frame Calc Time 0.48 [s]

* Additional info : at workstation (i9-9900K 3.6GHz, Nvidia GeForce RTX 2080Ti)
```
GPU Sonar Computation Time 0/100 [s]
CPU Sonar Sum - Correction 261/100 [s]
GPU FFT  Calculation  Time 0/100 [s]
(TOTAL) GPU Sonar Frame Calc Time 0.26 [s]
```

![RayRangeReduced](https://user-images.githubusercontent.com/7955120/98906567-a269c300-2500-11eb-923d-71ec7aa21f92.png)

woensug-choi committed:
sonar calculation almost done

woensug-choi commented:
@dtdavi1 @bsb808 Attempting to pass the Ubuntu CI Static code checking is taking too much time (uploading, wait for it to be done, fix accordingly, and iterate). Are there any methods that I check the code locally?

woensug-choi committed:
Summation gpu progress 50% reached 5 Hz

woensug-choi committed:
All gpu calc

woensug-choi commented:
### Finally, all calculations are done using GPU cuda cores.
- Some bottlenecks are observed when copying `nBeams x nRays x nFreq` (large array) between CPU and GPU in core calculation. About 0.15 seconds/frame at full range/rays calculations 
- Ray summation is not working as fast as others. It performs inner loop for column summation for `nFreq x nRays` matrix for each beams.
- Currently, fixed range/ray and variable refresh rate are used. This can be modified to fixed refresh rate and variable range/ray if needed.
- Not much difference for range/ray reduced calculations in sonartank settings.

### Benchmark results
-  at workstation (i9-9900K 3.6GHz, Nvidia GeForce RTX 2080Ti)
- 10 Hz achivable with Ray and Range reduced

|   512 Beams  |  Refresh  <br>  Rate [Hz]    | Total  <br>  Time [s]|  Core | Sum | Corr | FFT  |
|-------------------|--------------|-----------|------------|-----|-----|-----|
| **Full Calculation** <br> 60 m Range <br> 114 Rays      | **0.5 Hz** | 1.7 s | 0.3 | 1.26 | 0.05 | 0.03 |
| **Ray Reduced**  <br> 60 m Range <br> 11 Rays          | **3.0 Hz** | 0.27 s | 0.02 | 0.16 | 0.05 | 0.03 |
| **Ray/Range Reduced**  <br> 10 m Range <br> 11 Rays | **10 Hz** | 0.06 s | 0.00 | 0.04 | 0.01 | 0.00 |

![image](https://user-images.githubusercontent.com/7955120/99472466-015d8b00-298c-11eb-8844-83f2e859cdee.png)

woensug-choi committed:
static code checking and local tool

woensug-choi committed:
Update ci.yml

docker container image with nvidia cuda library

woensug-choi committed:
Update ci.yml

woensug-choi committed:
minor fix

woensug-choi committed:
Merge branch 'gpu_sonar_re' of github.com:Field-Robotics-Lab/nps_uw_sensors_gazebo into gpu_sonar_re

woensug-choi committed:
minor fix for static code checking

woensug-choi committed:
Update ci.yml

update Ubuntu CI test job name

woensug-choi commented:
@dtdavi1 I was trying to get it pass CI check on github for current commit. Editing ci.yml to use custom docker image including nvidia hub was successful. One last problem stays from the plugin 'shell' layer with following. 
```
./src/gazebo_ros_image_sonar.cpp:433:  Using C-style cast.  Use reinterpret_cast<float*>(...) instead  [readability/casting] [4]
```
But if I changed to `(float*)` part to `reinterpret_cast<float*>`, compile error occurs with `catkin_make`. Any easy ways to fix this?

dtdavi1 commented:
@woensug-choi It seems to be working on my computer now without errors.  I didn't look into the model file to check the plugin parameters, but the demo is running at about 10 Hz on my computer (reported sonar frame calc time is 7/100 to 9/100 seconds).

woensug-choi committed:
tabs format change check

woensug-choi committed:
fixed const_cast<float*>(reinterpret_cast<const float*>

woensug-choi committed:
fixed const_cast<float*>(reinterpret_cast<const float*>

woensug-choi committed:
fixed const_cast<float*>(reinterpret_cast<const float*>

woensug-choi commented:
> @woensug-choi It seems to be working on my computer now without errors. I didn't look into the model file to check the plugin parameters, but the demo is running at about 10 Hz on my computer (reported sonar frame calc time is 7/100 to 9/100 seconds).

Great! You've tested on host install environment right? Could you try docker as part of the PR review too? It's at `docker` folder in this repository. I'm wrapping up how to implement the target strength (=variable reflectivity). How difficult would it be to obtain and convey the material name or some sort to the sonar function?

woensug-choi commented:
### Issue on distance obtained from depth camera plugin 
#### Observed problem
1. Curvature at outer wall (wall behind the cylinder) on the final images above ([example image(latest)](https://user-images.githubusercontent.com/7955120/98906567-a269c300-2500-11eb-923d-71ec7aa21f92.png))
2. Point cloud obtained from 2D image of depth camera plugin(Figure left) shows unique unrealistic characteristic. 
   - (Figure left) shows Scatter3 plot of row, height, distance of image received at GPU sonar function.
   - Outer wall distance is flat while side walls are curved. This is strange, it should be all together straight lines or curved. Something different from previous wiki (Figure right, [Dave Wiki (image sonar, fls_model_standalone)](https://github.com/Field-Robotics-Lab/dave/wiki/fls_model_standalone)
![image](https://user-images.githubusercontent.com/7955120/100063290-57db3580-2e74-11eb-8c1e-5ca698bbc68e.png)

#### Possible solutions
As we were discussing to use ray tracing 'shell' sometime later, I think it should better be dealt at 'shell' side.
    - Such problem would not happen at ray tracing methods

@dtdavi1 Sorry that I've caught this problem much after the first 'shell' commit. I was having a discussion with Derek to find that outer wall curvature is something to be resolved in priority. Could you take a look if you can make it all straight lines or make curved ray angle dependent distances from the shell side?

dtdavi1 commented:
@woensug-choi I probably won't have much time to put into it this week, but I'll take a look when I can (it could be some sort of depth camera sensor artifact).  What did you use to generate the 3D plot?

woensug-choi commented:
> @woensug-choi I probably won't have much time to put into it this week, but I'll take a look when I can (it could be some sort of depth camera sensor artifact). What did you use to generate the 3D plot?

I printed (row, height, distance) on console terminal, saved it as csv file at excel, read in MATLAB to plot.

Data was passed as follows. 
1. `NpsGazeboRosImageSonar::ComputeSonarImage(const float *_src)`
2. `cv::Mat depth_image(this->height, this->width, CV_32FC1, const_cast<float*>(reinterpret_cast<const float*>(_src)));`
3. `NpsGazeboSonar::sonar_calculation_wrapper(depth_image, ...)`
4. `sonar_calculation<<<grid, block>>>(... ,d_depth_image, ...)` -> printed on console. 
Have not done any manipulation to it. Passed it to sonar_calculation GPU kernel function as a input.


Maybe this part is needed ([ref](https://github.com/NickSadjoli/uuv_simulator/blob/ca3a248337e2e7fdbf3db4835f92fcc778a17ab9/uuv_sensor_plugins/uuv_sensor_ros_plugins/src/gazebo_ros_image_sonar.cpp#L536))? (make point cloud and recalculate distance from sonar position and put it back to cv::mat to pass on)
Running this block at the `NpsGazeboRosImageSonar::ComputeSonarImage` seems to reproduce point cloud correctly. 
```
  double fl = ((double)this->width) / (2.0 *tan(hFOV/2.0));
  // convert depth to point cloud
  int index = 0;
  float* toCopyFrom = (float*)_src;
  for (uint32_t j=0; j<this->height; j++)
  {
    double pAngle;
    if (this->height>1) pAngle = atan2( (double)j - 0.5*(double)(this->height-1), fl);
    else            pAngle = 0.0;

    for (uint32_t i=0; i<this->width; i++)
    {
      double yAngle;
      if (this->width>1) yAngle = atan2( (double)i - 0.5*(double)(this->width-1), fl);
      else            yAngle = 0.0;

      double depth = toCopyFrom[index++];

      std::cout << depth * tan(yAngle) << "," << depth * tan(pAngle) << "," << depth << std::endl;
    }
  }
```

woensug-choi commented:
@dtdavi1 How was your holiday? :) Hope you had a good time during COVID. 
I am planning to do some major changes to this PR including the followings,
- [x] Half-precision boost **(UPDATE: Dec:11, removed due to compile errors depending on host machines)**
- [x] New structure for sonar calculation codes to precalculate things outside `NpsGazeboRosImageSonar::ComputeSonarImage`
The second subject might mess up with version control of the codes. I will work on the first subject first.

I've added some comments on the curvature of the input distances in the previous comment ([link](https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo/pull/25#issuecomment-733555945)). Hope it helps. Meanwhile, if you find a way to know the hit object model name, it would be progress on variational target strengths.

dtdavi1 commented:
@woensug-choi I wasn't entirely sure what the point of [ref](https://github.com/NickSadjoli/uuv_simulator/blob/ca3a248337e2e7fdbf3db4835f92fcc778a17ab9/uuv_sensor_plugins/uuv_sensor_ros_plugins/src/gazebo_ros_image_sonar.cpp#L536) was, but but after looking more closely, it looks like the depth image isn't encoded the way I had assumed.  Based on the formulas, the depth image values amount to a Z coordinate in the camera frame.  The azimuth and elevation calculations that I suggested don't look right either (if they were, the pAngle and yAngle calculations would be different).

What I still can't get my head around is why my misinterpretation would lead to the phenomenon that you're seeing.  If you interpret the image value as a range from the camera rather than a Z coordinate, the misinterpretation would result in shorter ranges than you would get using the point cloud?

Any thoughts?  We can add the point cloud calculations back into the plugin and work backwards from there as you suggest if you still think it will help.

woensug-choi commented:
@dtdavi1 I tried to figure out why but without any success at a short time glance. Considering the timeline, adding back the point cloud calculation and do backward calculation would be a workaround we can assure with neglectable calculation time increase. It would be great if the recalculated data is put back in cv::mat :)

woensug-choi committed:
SonarImag.msg published with topic name 'sonar_image'

woensug-choi commented:
@bsb808 @seankrag I've added ROS msg publisher :+1: 
It's using `sonar_msg` branch of `frl_msg` repository (https://github.com/Field-Robotics-Lab/frl_msgs/tree/sonar_msg)
Run command : `roslaunch nps_uw_sensors_gazebo sonar_tank_blueview_p900.launch verbose:=true`
ROS topic name : `/blueview_p900/sonar_image`
@seankrag You may try it without any further settings on the teamviewer remote computer that I've shared (email).

dtdavi1 committed:
Added point cloud calculation back into NpsGazeboRosImageSonar

The point cloud is only published if there is a subscriber to the ROS
topic, but it is computed if the sensor is active for use in the
sonar calculation as required.  Point cloud results are stored in the
PointCloud2 message and LOS range from the sensor to the point is also
stored in a cv::Mat image (CV_32FC1) as well.

dtdavi1 commented:
@woensug-choi I reintroduced the point cloud calculations into the plugin.  The point cloud message is being published to the point_cloud topic to support troubleshooting (it looks correct), and the full 3-channel point-cloud information is kept there.  I also added a cv::Mat point_cloud_image_ member variable to the class to store the line-of-sight range from the sensor to the points.  It's currently a single-channel 32-bit matrix (CV_32FC1), but it should be easy enough to change if you were hoping for a different encoding.  I also added inline Elevation() and Azimuth() methods to the class if you want to use those instead of stepping through angles as you work through the image.

woensug-choi committed:
sonar calculation distance fixed

woensug-choi commented:
@dtdavi1 Thank you for the quick update! I was able to fix the problem just fine! (see the attached screenshot)
Could you help me with these two other things?
1. Obtaining the target object name for variable target reflectivity  **(UPDATE @ Dec 10: Not doing)**
    It seems https://github.com/osrf/gazebo/blob/gazebo11/gazebo/rendering/UserCamera.hh#L185 is something we can adopt. Mabel said it should be sharing the same pose coordinates.
2. Adding library for the half-precision boost **(UPDATE @ Dec 7: Solved)**
    It's been consuming my time for too long. I was able to declare `__half` variables and also could use `__hadd` function to add them. But I was not able to use `__hmul` (multiplier) with compile error saying it's not been declared. There were forums that I should set linking with the higher library version ([link](https://github.com/NVIDIA/nccl/issues/9)). But I think I have the highest version CUDA version 11.0. I was trying this and that to add some possible FLAGS at CMakeLists but without any success.
![image](https://user-images.githubusercontent.com/7955120/101358630-77e00f80-38de-11eb-8f33-0e65199eb8b3.png)

bsb808 commented on src/gazebo_ros_image_sonar.cpp:
```diff
@@ -0,0 +1,855 @@
+/*
```
Are these copyright notices still appropriate?

woensug-choi replied:
I shall erase them in the next commit

bsb808 commented:
@woensug-choi 
It looks like with commit 86d7391911d00e5827a2fd0b82ba68e79d48667a you added the sonar image ROS message type.  

I think we may want to avoid copying the message definition into our own repository.  Instead, we should consider introducing the [imaging_sonar_msgs](https://gitlab.com/apl-ocean-engineering/imaging_sonar_msgs) package as a dependehttps://gitlab.com/apl-ocean-engineering/imaging_sonar_msgsncy.  This would avoid divergence of two very similar "standard" messages.

woensug-choi commented:
> @woensug-choi
> It looks like with commit [86d7391](https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo/commit/86d7391911d00e5827a2fd0b82ba68e79d48667a) you added the sonar image ROS message type.
> 
> I think we may want to avoid copying the message definition into our own repository. Instead, we should consider introducing the [imaging_sonar_msgs](https://gitlab.com/apl-ocean-engineering/imaging_sonar_msgs) package as a dependehttps://gitlab.com/apl-ocean-engineering/imaging_sonar_msgsncy. This would avoid divergence of two very similar "standard" messages.

Thank you for the comments! Are there ways to add a dependent repository other than adding it as a submodule? I remember the EPIC glider plugin's previous comment that adding a submodule repository is not desirable before.

bsb808 commented:
> > @woensug-choi
> > It looks like with commit [86d7391](https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo/commit/86d7391911d00e5827a2fd0b82ba68e79d48667a) you added the sonar image ROS message type.
> > I think we may want to avoid copying the message definition into our own repository. Instead, we should consider introducing the [imaging_sonar_msgs](https://gitlab.com/apl-ocean-engineering/imaging_sonar_msgs) package as a dependehttps://gitlab.com/apl-ocean-engineering/imaging_sonar_msgsncy. This would avoid divergence of two very similar "standard" messages.
> 
> Thank you for the comments! Are there ways to add a dependent repository other than adding it as a submodule? I remember the EPIC glider plugin's previous comment that adding a submodule repository is not desirable before.

Yes - I added a PR on your branch to illustrate: https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo/pull/31

woensug-choi committed:
edited copyright and sonarimage.msg dependency

woensug-choi commented:
> > > @woensug-choi
> > > It looks like with commit [86d7391](https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo/commit/86d7391911d00e5827a2fd0b82ba68e79d48667a) you added the sonar image ROS message type.
> > > I think we may want to avoid copying the message definition into our own repository. Instead, we should consider introducing the [imaging_sonar_msgs](https://gitlab.com/apl-ocean-engineering/imaging_sonar_msgs) package as a dependehttps://gitlab.com/apl-ocean-engineering/imaging_sonar_msgsncy. This would avoid divergence of two very similar "standard" messages.
> > 
> > 
> > Thank you for the comments! Are there ways to add a dependent repository other than adding it as a submodule? I remember the EPIC glider plugin's previous comment that adding a submodule repository is not desirable before.
> 
> Yes - I added a PR on your branch to illustrate: #31

Thank you! Added!

woensug-choi committed:
introducing half precision calculation and int summation but no increase in computation time

woensug-choi committed:
fix static code checking errors

woensug-choi committed:
add gitlab imaging_sonar_msgs to ci

woensug-choi committed:
ci image_sonar_msgs with url path

woensug-choi commented:
I've been trying to get the final image through rqt window. I was not able to commit since the remote computer is down.

woensug-choi committed:
standalone bundling including rqt viewer

woensug-choi commented:
@crvogt 
This part converts cv::mat of final sonar data (nRow = nBeams, nCols = nRangeVector) 
https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo/blob/1f72f090f5ef06cb5be1f6a96734e1e0d5a01fcc/src/gazebo_ros_image_sonar.cpp#L683-L755

What it does is
1. Allocate data to a new array with 4*nBeams that will be mapped 360 degrees. (Every other part is empty but only places of 2.5*nbeams~3.5*nBeams (total of nBeams rows) are filled up to have the final image to be located at the upper circle.
2. Rescale
3. Polar coordinate transform using linearPolar
4. Search bounding box for crop range
5. Crop image

What I get is something like this. Which the walls are not expected to be curvy. 
Run command : `roslaunch nps_uw_sensors_gazebo sonar_tank_blueview_p900.launch verbose:=true`

![image](https://user-images.githubusercontent.com/7955120/101931323-99156880-3c1c-11eb-8185-99e88bbe88b5.png)

woensug-choi commented:
@dtdavi1 
At the final commit, I've rearranged the final rqt plot windows. I could not find a way to remove the empty area below the 1 row, 3 columns plot. Where should I edit to fit the window?
![image](https://user-images.githubusercontent.com/7955120/101931392-afbbbf80-3c1c-11eb-8c23-0baf2c163a19.png)

woensug-choi commented:
The last commits from two different machines messed things up at the very end. It's not calculating right...

woensug-choi commented:
@dtdavi1 @crvogt  @seankrag 
The last commits from two different machines messed things up at the very end. Please run at commit `4502ef805d7ff1a4d607d48d2376c684d6dc3aed` -> made as new branch `gpu_sonar_review`
```
 git checkout 4502ef805d7ff1a4d607d48d2376c684d6dc3aed
```
I'll make another branch to do the formal review early next week.
To-do (from branch for review `gpu_sonar_review`)
- [x] Remove half-precision to avoid machine dependency
- [ ] Test docker image

crvogt commented:
@woensug-choi Nothing stands out to me as immediately out of place. I'm trying to build on my computer so I can try out a couple things, but having a little struggle with it.

dtdavi1 commented:
@crvogt When I set everything up on my machine I used a brute force approach--see where the compile breaks and then look up how to install whatever caused the failure.  In the end, I think I used apt-get to install cuda and cuda-nvcc-11-1 (but don't quote me on that--I had a few aborted attempts that I had to back out of before I resorted to brute force).

@woensug-choi Rather than checking out the commit for now, you might be able to do a soft reset to that commit, recommit everything on top of it correctly, and do a forced push back to the server.  Alternatively, you could create a tag at that commit (more straightforward checkout) and then fixing everything in that branch.  Either option might be a little cleaner than creating a new branch.

crvogt commented:
@dtdavi1 Thanks. I eventually managed to get through the the CUDA errors after editing my `.bashrc`. Currently looking for `imaging_sonar_msgs`. Should I be pulling a package from APL?

seankrag commented:
Carson,

Yes. The repo is at:
https://gitlab.com/apl-ocean-engineering/imaging_sonar_msgs

Sean

On Fri, Dec 11, 2020 at 2:16 PM crvogt <notifications@github.com> wrote:

> @dtdavi1 <https://github.com/dtdavi1> Thanks. I eventually managed to get
> through the the CUDA errors after editing my .bashrc. Currently looking
> for imaging_sonar_msgs. Should I be pulling a package from APL?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo/pull/25#issuecomment-743456025>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARJIVTZXPEX6FD7ZZX26PYLSUKK4ZANCNFSM4THJQXQA>
> .
>

crvogt commented:
@seankrag Thanks!

seankrag commented:
@dtdavi1 @crvogt

@woensug-choi , I have been testing the SonarImage messages against the UW:APL viewer. My process is to run the gpu_sonar on my remote machine, save the bag files, and play them back on my local machine.

I have been looking at messages from three branches

1. gpu_sonar_re_sonarmsg
2. gpu_sonar_re
3. gpu_sonar_review

Of these, the first one is the only bag file that **image_view** renders as expected (although the image is just noise); the other two, however, are completely black.
![frame0000](https://user-images.githubusercontent.com/72518351/101970139-8465b000-3bdd-11eb-8870-d82d8f40aeec.jpg)

I tested the UW:APL tools using one of their bag files as input, and everything works fine. So I suspect that we're stuffing the SonarImage message incorrectly.  Comparing our bag files to the UW:APL dataset, our messages differ in two obvious ways:

1. Our data_size field is set to 0, their data_size field is set to 1. A quick fix may be setting this field to 1.
2. We have a single value in the elevation angle array, they leave this field blank. According to the message definition:
> If left empty, source is assumed to be a 1D sonar with an elevation of { 0.0 }

Beyond that, I may need to look at array ordering, etc. I can create a new branch to start debugging the SonarImage output messages. Which branch would be best to start from?

woensug-choi committed:
recommiting after reset

woensug-choi committed:
Merge branch 'master' into gpu_sonar_re

woensug-choi commented:
@dtdavi1 @seankrag @crvogt
I've reset the commits on this branch. Use this branch (`gpu_sonar_re`) only from now on :+1: 
At the rqt_image_view of the final sonar image, set range 0.25 and colormap 'hot' to obtain the image below.
![image](https://user-images.githubusercontent.com/7955120/102041571-24584f00-3e13-11eb-9370-540f1af71d7b.png)

@dtdavi1 Thank you for the comment on GitHub soft reset. Made soft reset and force push on this branch. 

@seankrag It looks promising! just needs some iterations to match the format. Can you try it again now?
I've put `sizeof(float) * nFreq * nBeams` for `data_size`. Is their field has just int 1?, Left elevation array blank.
Linzy has shared a new repository for `SonarImage.msg`. What we had was the Gitlab one while the new shared repository is in Github (https://github.com/apl-ocean-engineering/acoustic_msgs). Do we need to change to Github one?

@crvogt Were you able to run it? Can you try it again? Also, it would be great if you can add some missing instructions on installation at https://github.com/Field-Robotics-Lab/dave/wiki/Multibeam-Forward-Looking-Sonar#Requirements

woensug-choi committed:
commit messes all fixed, half precision removed

woensug-choi committed:
Dockerfile cuda install version conflict fix

crvogt commented:
@woensug-choi I've managed to get it running. The quickstart launch command is eating a lot of RAM for me (>10GB per min). 

Commenting out `<node name="rqt_images" pkg="rqt_gui" type="rqt_gui" output="screen" args = "--perspective-file $(find nps_uw_sensors_gazebo)/config/sonar_tank_blueview_p900_images.perspective" />` and RAM usage is normal.

Is there an argument I need to pass or is it collecting data in a rosbag?

seankrag commented:
> Linzy has shared a new repository for `SonarImage.msg`. What we had was the Gitlab one while the new shared repository is in Github (https://github.com/apl-ocean-engineering/acoustic_msgs). Do we need to change to Github one?

@woensug-choi , the message format is identical. I suggest we hold off on switching to the Github repository for now until I finish testing the viewer.

Commit:
setting data_size = 1

woensug-choi commented:
> @woensug-choi I've managed to get it running. The quickstart launch command is eating a lot of RAM for me (>10GB per min).
> 
> Commenting out `<node name="rqt_images" pkg="rqt_gui" type="rqt_gui" output="screen" args = "--perspective-file $(find nps_uw_sensors_gazebo)/config/sonar_tank_blueview_p900_images.perspective" />` and RAM usage is normal.
> 
> Is there an argument I need to pass or is it collecting data in a rosbag?

I am not sure what could manage those large memory usages. But in practical use, the line you've commented out should be commented out. (`sonar_tank_blueview_p900_images.perspective` is showing `real_image`, `depth_image`, `normal_image` for debugging purposes). Commenting out would not hurt generating the sonar image. I will comment them out before the merge and leave notes on wikis how to enable them.

woensug-choi committed:
Update sonar_tank_blueview_p900.launch

Commenting out the `real_image`,`depth_image`,`normal_image` viewer for saving memory usage for the initial start-up settings.

Commit:
Swapped inner and outer loops to fill the SonarImage intensities array in azimuth-major order.
This change enables usage of UW:APL's libdraw_sonar tool to view sonar data.

Commit:
Merge branch 'gpu_sonar_re' of https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo into gpu_sonar_re

seankrag commented:
> @woensug-choi , I made two small changes to gazebo_ros_image_sonar.cpp for publishing SonarImage msgs:

1. set data_size=1 for consistency with uw-apl bag files; this field is the size of each datum in bytes
2. exchanged the order of for loops used to populate the intensities array in azimuth-major order

This allows me to run the uw-apl viewer, which converts an imaging_sonar_msgs/SonarImage into a sensor_msgs/Image for display by ROS image_view. I still need to investigate whether we can fine-tune the uw-apl viewer, but your OpenCV code currently produces a higher-resolution image. See comparison below.
**sonar_image**:
![sonar_image](https://user-images.githubusercontent.com/72518351/102174767-07dc0700-3e53-11eb-8fee-109136df5208.png)
**sonar_image_raw** rendered by viewer:
![sonar_image_raw_to_viewer](https://user-images.githubusercontent.com/72518351/102174773-09a5ca80-3e53-11eb-8254-b5a32f6fc8ff.png)

woensug-choi commented:
@seankrag Glad to see it showing the image! Thanks! Are there any methods to modify the color scale max/min?

@crvogt This looks like something we could reference! https://github.com/apl-ocean-engineering/libdraw_sonar/blob/c526cc2bde57cae0e598eaf891680d5a4fcd3e67/lib/DrawSonar.cpp

crvogt commented:
> @crvogt This looks like something we could reference! https://github.com/apl-ocean-engineering/libdraw_sonar/blob/c526cc2bde57cae0e598eaf891680d5a4fcd3e67/lib/DrawSonar.cpp

Haven't been able to get a sonar image at all yet. To even publish on the `sonar_image` topic I guess the code is looking for a subscriber to the `image_depth` topic before it begins to publish. 

However, when I subscribe to `image_depth`, that's when the RAM draw goes way up (writing to `/tmp`? I'll check). though the `sonar_image` message is finally published. However, I'm not seeing an actual image yet, just getting a black screen.

EDIT: Saw your note about setting to "Hot" and "0.25m". Getting an image now! (though, still with the RAM draw)

seankrag commented:
> @seankrag Glad to see it showing the image! Thanks! Are there any methods to modify the color scale max/min?

@woensug-choi I will check, but much of the viewer code settings appear baked-in. Separately, I followed up with Laura Lindzey, who commented on the difference between our 32-bit OpenCV image and the APL:UW viewer's 8-bit image (it can also support16-bit images). She is happy to extend their viewer to 32-bit images if we like, but all of their sonar produce only 8- or 16-bit intensities (and she wasn't aware of actual sonars that produce higher levels of precision). We already cast our data into 8-bit values for the SonarImage msg, so this may really just a design/philosophical question: @bsb808 Do we need/want the high-precision currently used by our sonar model? Could we realize performance benefits by using 16-bit instead of 32-bit?

woensug-choi committed:
New bundled viwer

woensug-choi committed:
Memory usage fix

woensug-choi commented:
#### Introducing new bundled viewer included in this PR (cv mapping fixed)
- **Only in monocolor**
- Color scalable with `<plotScaler>1</plotScaler>` at `model.sdf`
*** 
![image](https://user-images.githubusercontent.com/7955120/102305129-885d4d80-3fa2-11eb-88bc-33952636fd64.png)

@crvogt The reference I've found was pretty straight forward to implement! Thank you for noticing the RAM usage! I was using a machine with 64GB RAM... I've found some not freeing memory allocation on Cuda calculation! Now it should be fixed! Could you do try user experience all over once again? I've updated [Wiki page](https://github.com/Field-Robotics-Lab/dave/wiki/Multibeam-Forward-Looking-Sonar) too.

@seankrag I've made a new viewer bundled in this PR referencing UW APL's viewer codes. Currently, it is using the 16-bit image msg (OpenCV's CV_16UC1 sent as msgs MONO16). I was getting exactly the same image of your post with 8-bit data (CV_8UC1 OpenCV format; MONO8 msg format). In summary, I think 16-bit looks good enough. The only advantage of having 32-bit float data on my side is that it's easy to colorize with ROS bundled `rqt_image_view` viewer package. It seems with MONO6 or MONO16 msg format, the color scale could not be applied.

@dtdavi1 I believe this PR is now finally ready to go for the official review! We could use the last iteration to fix things if any problems are found before our celebrating meeting.

dtdavi1 committed:
Cosmetic changes to fix Git static code check errors

dtdavi1 reviewed (approved):
Everything seems to run correctly and the images are all right.  I am seeing 2 "Node::Advertis(): Error advertising a topic. Did you forget to start the discovery service?" messages, but I don't know what they are associated with.

I added one commit to fix the static code checks.  A different check is still failing because of an error associated with finding the APL message repo.

woensug-choi commented:
> Everything seems to run correctly and the images are all right. I am seeing 2 "Node::Advertis(): Error advertising a topic. Did you forget to start the discovery service?" messages, but I don't know what they are associated with.
> 
> I added one commit to fix the static code checks. A different check is still failing because of an error associated with finding the APL message repo.

Thank you for the quick review!
- "Node::Advertis(): Error
  - Where can I see the msg? I don't see anything like it on the terminal screen. I am moving this to an Issue page.
- CI check missing APL message repo
  - I've been trying to get this working for a while. But without success. This part is added to [workflow file](https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo/actions/runs/426921806/workflow) fix it. But did not work. Do you know any fixes to this?
   ```yml
      - name: Checkout UW imaging_sonar_msgs
        uses: actions/checkout@v2
        with:
          url: https://gitlab.com/apl-ocean-engineering/imaging_sonar_msgs.git
          path: imaging_sonar_msgs
   ```
(UPDATE)
It seems Github workflow is not allowed to use a repository outside Github. Changed SonarImage.msg repository to the newly made one at the Github. But still, the ci check fails. Since its not something to do with the functionality I will merge as it is for now. 
[New workflow file with Github repository](https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo/actions/runs/427248728/workflow)

woensug-choi committed:
Update ci.yml

Changing SonarImage.msg source from Gitlab to Github

woensug-choi committed:
Changing SonarImage.msg from Gitlab to Github

woensug-choi committed:
Merge branch 'gpu_sonar_re' of github.com:Field-Robotics-Lab/nps_uw_sensors_gazebo into gpu_sonar_re
